File: backend/app/routers/requests.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..audit import log_action
from ..database import get_db
from ..deps import get_current_user, require_roles
from ..models import (
    Guard,
    GuardStatus,
    MinusOneRequest,
    RequestStatus,
    Role,
    User,
    utcnow,
)
from ..schemas import MinusOneCreate, RejectBody, RequestOut
from ..ws import notify

router = APIRouter(prefix="/api/requests", tags=["requests"])


def serialize(req: MinusOneRequest, db: Session) -> RequestOut:
    guard = db.get(Guard, req.guard_id)
    out = RequestOut.model_validate(req)
    if guard:
        out.guard_name = guard.name
        out.tower_id = guard.tower_id
    return out


@router.get("", response_model=list[RequestOut])
def list_requests(
    status: RequestStatus | None = None,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    q = db.query(MinusOneRequest)
    if status:
        q = q.filter(MinusOneRequest.status == status)
    # Wachgänger sehen nur eigene Anfragen.
    if user.role == Role.WACHGAENGER:
        q = q.filter(MinusOneRequest.requested_by == user.id)
    elif user.role == Role.TURMFUEHRER:
        tower_guard_ids = [
            g.id for g in db.query(Guard).filter(Guard.tower_id == user.tower_id).all()
        ]
        q = q.filter(MinusOneRequest.guard_id.in_(tower_guard_ids or [-1]))
    reqs = q.order_by(MinusOneRequest.created_at.desc()).all()
    return [serialize(r, db) for r in reqs]


@router.post("/minus-one", response_model=RequestOut, status_code=201)
def request_minus_one(
    body: MinusOneCreate,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    guard = db.get(Guard, body.guard_id)
    if not guard:
        raise HTTPException(404, "Wachgänger nicht gefunden")
    # Wachgänger darf nur für sich selbst beantragen.
    if user.role == Role.WACHGAENGER and guard.user_id != user.id:
        raise HTTPException(403, "Nur für eigenen Wachgänger möglich")

    existing = (
        db.query(MinusOneRequest)
        .filter(
            MinusOneRequest.guard_id == guard.id,
            MinusOneRequest.status.in_([RequestStatus.PENDING, RequestStatus.APPROVED]),
        )
        .first()
    )
    if existing:
        raise HTTPException(409, "Es existiert bereits eine offene/aktive -1 Anfrage")

    req = MinusOneRequest(
        guard_id=guard.id,
        requested_by=user.id,
        reason=body.reason,
        note=body.note,
        status=RequestStatus.PENDING,
    )
    db.add(req)
    db.flush()
    log_action(db, user, "MINUS_ONE_REQUESTED", "request", req.id,
               {"guard": guard.name, "reason": body.reason.value})
    db.commit()
    db.refresh(req)
    notify("requests_changed", {"id": req.id, "status": "PENDING"})
    return serialize(req, db)


@router.post("/{request_id}/approve", response_model=RequestOut)
def approve(
    request_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(require_roles(Role.HAUPTWACHE)),
):
    req = db.get(MinusOneRequest, request_id)
    if not req:
        raise HTTPException(404, "Anfrage nicht gefunden")
    if req.status != RequestStatus.PENDING:
        raise HTTPException(409, "Anfrage ist nicht mehr offen")

    req.status = RequestStatus.APPROVED
    req.decided_at = utcnow()
    req.decided_by = user.id
    guard = db.get(Guard, req.guard_id)
    if guard:
        guard.status = GuardStatus.MINUS_ONE
    log_action(db, user, "MINUS_ONE_APPROVED", "request", req.id)
    db.commit()
    db.refresh(req)
    notify("requests_changed", {"id": req.id, "status": "APPROVED"})
    notify("guards_changed", {"id": req.guard_id})
    return serialize(req, db)


@router.post("/{request_id}/reject", response_model=RequestOut)
def reject(
    request_id: int,
    body: RejectBody,
    db: Session = Depends(get_db),
    user: User = Depends(require_roles(Role.HAUPTWACHE)),
):
    req = db.get(MinusOneRequest, request_id)
    if not req:
        raise HTTPException(404, "Anfrage nicht gefunden")
    if req.status != RequestStatus.PENDING:
        raise HTTPException(409, "Anfrage ist nicht mehr offen")

    req.status = RequestStatus.REJECTED
    req.decided_at = utcnow()
    req.decided_by = user.id
    req.rejection_reason = body.rejection_reason
    log_action(db, user, "MINUS_ONE_REJECTED", "request", req.id,
               {"reason": body.rejection_reason})
    db.commit()
    db.refresh(req)
    notify("requests_changed", {"id": req.id, "status": "REJECTED"})
    return serialize(req, db)


@router.post("/{request_id}/return", response_model=RequestOut)
def return_to_area(
    request_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(get_current_user),
):
    req = db.get(MinusOneRequest, request_id)
    if not req:
        raise HTTPException(404, "Anfrage nicht gefunden")
    if user.role == Role.WACHGAENGER and req.requested_by != user.id:
        raise HTTPException(403, "Nur eigene Rückkehr meldbar")
    if req.status != RequestStatus.APPROVED:
        raise HTTPException(409, "Nur eine aktive -1 kann zurückgemeldet werden")

    req.status = RequestStatus.RETURNED
    req.returned_at = utcnow()
    guard = db.get(Guard, req.guard_id)
    if guard:
        guard.status = GuardStatus.IN_AREA
    log_action(db, user, "MINUS_ONE_RETURNED", "request", req.id)
    db.commit()
    db.refresh(req)
    notify("requests_changed", {"id": req.id, "status": "RETURNED"})
    notify("guards_changed", {"id": req.guard_id})
    return serialize(req, db)

File: apps/api/src/routes/requests.ts
import type { FastifyInstance } from 'fastify';
import { and, eq, inArray, sql } from 'drizzle-orm';
import {
  minusOneSchema,
  kFahrtRequestSchema,
  rejectSchema,
  parsePositiveInt,
  type RequestView,
  type RequestStatus,
  type RequestKind,
  type Reason
} from '@turmstatus/shared';
import { minusOneRequests, guards, towers, users } from '../db/schema.js';
import { requireAuth, requireWachfuehrer, requireRole } from '../plugins/auth.js';
import { parseBody } from '../lib/validate.js';
import { recordAudit } from '../lib/audit.js';
import type { ViewScope } from '../types/fastify.js';

function listRequests(app: FastifyInstance, scope: ViewScope, status?: string): RequestView[] {
  const conds = [];
  if (!scope.all) conds.push(eq(guards.ownerId, scope.scopeId));
  if (status) conds.push(eq(minusOneRequests.status, status));
  const where = conds.length ? and(...conds) : undefined;

  const rows = app.db
    .select({
      r: minusOneRequests,
      guardName: guards.name,
      guardOwner: guards.ownerId,
      towerId: guards.towerId,
      towerName: towers.name,
      requesterName: users.fullName,
      requesterUser: users.username
    })
    .from(minusOneRequests)
    .leftJoin(guards, eq(minusOneRequests.guardId, guards.id))
    .leftJoin(towers, eq(guards.towerId, towers.id))
    .leftJoin(users, eq(minusOneRequests.requestedBy, users.id))
    .where(where)
    .orderBy(sql`${minusOneRequests.createdAt} DESC`)
    .all();

  return rows.map((row): RequestView => ({
    id: row.r.id,
    guardId: row.r.guardId,
    guardName: row.guardName,
    towerId: row.towerId,
    towerName: row.towerName,
    requestedBy: row.r.requestedBy,
    requestedByName: row.requesterName ?? row.requesterUser,
    kind: (row.r.kind ?? 'MINUS_ONE') as RequestKind,
    reason: (row.r.reason ?? null) as Reason | null,
    note: row.r.note,
    status: row.r.status as RequestStatus,
    rejectionReason: row.r.rejectionReason,
    createdAt: row.r.createdAt,
    decidedAt: row.r.decidedAt,
    decidedBy: row.r.decidedBy,
    returnedAt: row.r.returnedAt,
    ownerId: row.guardOwner
  }));
}

export async function requestRoutes(app: FastifyInstance): Promise<void> {
  app.get('/api/requests', { preHandler: requireAuth }, async (req) => {
    const status = (req.query as { status?: string }).status;
    return listRequests(app, req.scope, status);
  });

  // -1 beantragen.
  app.post(
    '/api/requests/minus-one',
    { preHandler: [requireAuth, requireRole('WACHGAENGER', 'BOOTSFUEHRER', 'WACHFUEHRER')] },
    async (req, reply) => {
      const body = parseBody(minusOneSchema, req.body, reply);
      if (!body) return;
      const scope = req.scope;
      const guard = app.db.select().from(guards).where(eq(guards.id, body.guardId)).get();
      const inScope = guard && (scope.all || guard.ownerId === scope.scopeId);
      if (!guard || !inScope) return reply.code(404).send({ error: 'Wachgänger nicht gefunden' });

      const open = app.db
        .select({ id: minusOneRequests.id })
        .from(minusOneRequests)
        .where(and(eq(minusOneRequests.guardId, body.guardId), inArray(minusOneRequests.status, ['PENDING', 'APPROVED'])))
        .get();
      if (open) return reply.code(409).send({ error: 'Es gibt bereits eine offene/aktive -1-Anfrage' });

      const row = app.db
        .insert(minusOneRequests)
        .values({
          guardId: body.guardId,
          requestedBy: req.session.user!.id,
          reason: body.reason,
          note: body.note ?? null,
          status: 'PENDING'
        })
        .returning({ id: minusOneRequests.id })
        .get();
      recordAudit(req, 'request.minus-one', 'request', row.id, { guardId: body.guardId, reason: body.reason });
      app.realtime.broadcast('requests-updated');
      return reply.code(201).send({ id: row.id });
    }
  );

  // K-Fahrt (Kontrollfahrt) beantragen – Bootsführer (oder Wachführer).
  app.post(
    '/api/requests/k-fahrt',
    { preHandler: [requireAuth, requireRole('BOOTSFUEHRER', 'WACHFUEHRER')] },
    async (req, reply) => {
      const body = parseBody(kFahrtRequestSchema, req.body, reply);
      if (!body) return;
      const scope = req.scope;
      const guard = app.db.select().from(guards).where(eq(guards.id, body.guardId)).get();
      const inScope = guard && (scope.all || guard.ownerId === scope.scopeId);
      if (!guard || !inScope) return reply.code(404).send({ error: 'Wachgänger nicht gefunden' });

      const open = app.db
        .select({ id: minusOneRequests.id })
        .from(minusOneRequests)
        .where(
          and(
            eq(minusOneRequests.guardId, body.guardId),
            eq(minusOneRequests.kind, 'K_FAHRT'),
            inArray(minusOneRequests.status, ['PENDING', 'APPROVED'])
          )
        )
        .get();
      if (open) return reply.code(409).send({ error: 'Es gibt bereits eine offene/aktive K-Fahrt-Anfrage' });

      const row = app.db
        .insert(minusOneRequests)
        .values({
          guardId: body.guardId,
          requestedBy: req.session.user!.id,
          kind: 'K_FAHRT',
          reason: null,
          note: body.note ?? null,
          status: 'PENDING'
        })
        .returning({ id: minusOneRequests.id })
        .get();
      recordAudit(req, 'request.k-fahrt', 'request', row.id, { guardId: body.guardId });
      app.realtime.broadcast('requests-updated');
      return reply.code(201).send({ id: row.id });
    }
  );

  // Genehmigen – nur der Owner-Wachführer des Wachgängers.
  app.post('/api/requests/:id/approve', { preHandler: [requireAuth, requireWachfuehrer] }, async (req, reply) => {
    const id = parsePositiveInt((req.params as { id: string }).id);
    if (!id) return reply.code(400).send({ error: 'Ungültige ID' });
    const uid = req.session.user!.id;
    const found = app.db
      .select({ r: minusOneRequests, guardOwner: guards.ownerId, guardId: guards.id })
      .from(minusOneRequests)
      .leftJoin(guards, eq(minusOneRequests.guardId, guards.id))
      .where(eq(minusOneRequests.id, id))
      .get();
    if (!found || found.guardOwner !== uid) return reply.code(404).send({ error: 'Anfrage nicht gefunden' });
    if (found.r.kind === 'K_FAHRT')
      return reply.code(409).send({ error: 'Kontrollfahrten werden über „K-Fahrt setzen" aktiviert' });
    if (found.r.status !== 'PENDING') return reply.code(409).send({ error: 'Anfrage ist nicht mehr offen' });

    app.db
      .update(minusOneRequests)
      .set({ status: 'APPROVED', decidedAt: sql`CURRENT_TIMESTAMP`, decidedBy: uid })
      .where(eq(minusOneRequests.id, id))
      .run();
    app.db.update(guards).set({ status: 'MINUS_ONE', updatedAt: sql`CURRENT_TIMESTAMP` }).where(eq(guards.id, found.r.guardId)).run();
    recordAudit(req, 'request.approve', 'request', id);
    app.realtime.broadcast('requests-updated');
    app.realtime.broadcast('guards-updated');
    app.realtime.broadcast('towers-updated');
    return { ok: true };
  });

  // K-Fahrt setzen – nur der Owner-Wachführer. Bewusst NICHT über „genehmigen":
  // erst wenn der Wachführer die K-Fahrt aktiv setzt, wird der Turm um 2 WG reduziert.
  app.post('/api/requests/:id/set-k-fahrt', { preHandler: [requireAuth, requireWachfuehrer] }, async (req, reply) => {
    const id = parsePositiveInt((req.params as { id: string }).id);
    if (!id) return reply.code(400).send({ error: 'Ungültige ID' });
    const uid = req.session.user!.id;
    const found = app.db
      .select({ r: minusOneRequests, guardOwner: guards.ownerId })
      .from(minusOneRequests)
      .leftJoin(guards, eq(minusOneRequests.guardId, guards.id))
      .where(eq(minusOneRequests.id, id))
      .get();
    if (!found || found.guardOwner !== uid) return reply.code(404).send({ error: 'Anfrage nicht gefunden' });
    if (found.r.kind !== 'K_FAHRT') return reply.code(409).send({ error: 'Keine Kontrollfahrt-Anfrage' });
    if (found.r.status !== 'PENDING') return reply.code(409).send({ error: 'Anfrage ist nicht mehr offen' });

    app.db
      .update(minusOneRequests)
      .set({ status: 'APPROVED', decidedAt: sql`CURRENT_TIMESTAMP`, decidedBy: uid })
      .where(eq(minusOneRequests.id, id))
      .run();
    recordAudit(req, 'request.set-k-fahrt', 'request', id);
    app.realtime.broadcast('requests-updated');
    app.realtime.broadcast('towers-updated');
    return { ok: true };
  });

  // Ablehnen – nur der Owner-Wachführer.
  app.post('/api/requests/:id/reject', { preHandler: [requireAuth, requireWachfuehrer] }, async (req, reply) => {
    const id = parsePositiveInt((req.params as { id: string }).id);
    if (!id) return reply.code(400).send({ error: 'Ungültige ID' });
    const body = parseBody(rejectSchema, req.body, reply);
    if (!body) return;
    const uid = req.session.user!.id;
    const found = app.db
      .select({ r: minusOneRequests, guardOwner: guards.ownerId })
      .from(minusOneRequests)
      .leftJoin(guards, eq(minusOneRequests.guardId, guards.id))
      .where(eq(minusOneRequests.id, id))
      .get();
    if (!found || found.guardOwner !== uid) return reply.code(404).send({ error: 'Anfrage nicht gefunden' });
    if (found.r.status !== 'PENDING') return reply.code(409).send({ error: 'Anfrage ist nicht mehr offen' });

    app.db
      .update(minusOneRequests)
      .set({ status: 'REJECTED', decidedAt: sql`CURRENT_TIMESTAMP`, decidedBy: uid, rejectionReason: body.rejectionReason ?? null })
      .where(eq(minusOneRequests.id, id))
      .run();
    recordAudit(req, 'request.reject', 'request', id);
    app.realtime.broadcast('requests-updated');
    return { ok: true };
  });

  // +1 / Rückkehr – Owner-Wachführer ODER jemand im selben Mandanten-Scope.
  app.post('/api/requests/:id/return', { preHandler: requireAuth }, async (req, reply) => {
    const id = parsePositiveInt((req.params as { id: string }).id);
    if (!id) return reply.code(400).send({ error: 'Ungültige ID' });
    const scope = req.scope;
    const found = app.db
      .select({ r: minusOneRequests, guardOwner: guards.ownerId, guardId: guards.id })
      .from(minusOneRequests)
      .leftJoin(guards, eq(minusOneRequests.guardId, guards.id))
      .where(eq(minusOneRequests.id, id))
      .get();
    const inScope = found && (scope.all || found.guardOwner === scope.scopeId);
    if (!found || !inScope) return reply.code(404).send({ error: 'Anfrage nicht gefunden' });
    if (found.r.status !== 'APPROVED')
      return reply.code(409).send({ error: 'Nur genehmigte -1 / aktive K-Fahrten können zurückgemeldet werden' });

    app.db
      .update(minusOneRequests)
      .set({ status: 'RETURNED', returnedAt: sql`CURRENT_TIMESTAMP` })
      .where(eq(minusOneRequests.id, id))
      .run();
    // Bei -1 kehrt der Wachgänger in den Bereich zurück; die K-Fahrt hat den Guard-Status nicht verändert.
    if (found.r.kind !== 'K_FAHRT') {
      app.db.update(guards).set({ status: 'IN_AREA', updatedAt: sql`CURRENT_TIMESTAMP` }).where(eq(guards.id, found.r.guardId)).run();
      app.realtime.broadcast('guards-updated');
    }
    recordAudit(req, 'request.return', 'request', id);
    app.realtime.broadcast('requests-updated');
    app.realtime.broadcast('towers-updated');
    return { ok: true };
  });
}
